### PR TITLE
Basic SMP support

### PR DESCRIPTION
**Pull request by estutzenberger**
* Supports Secure Connections (needs more testing)
* Supports Legacy security as well
* Still need to add tests

### CODE REVIEW

File: linux/hci/conn.go [hunk 1]
```diff
@@ -184,6 +201,81 @@ func (c *Conn) Write(sdu []byte) (int, error) {
 	return sent, nil
 }
 
+func (c *Conn) initPairingContext() {
+	smp := c.smp
+
+	la := c.LocalAddr().Bytes()
+	//assume la is public address for now
+	//todo: figure out a better way
```
Comment (donaldschen) [inline linux/hci/conn.go]: couldn't we detect public vs random by checking the MSB of the local addr?

File: linux/hci/hci.go [hunk 2]
```diff
@@ -128,12 +132,13 @@ func (h *HCI) Init() error {
...
 	h.evth[evt.DisconnectionCompleteCode] = h.handleDisconnectionComplete
 	h.evth[evt.NumberOfCompletedPacketsCode] = h.handleNumberOfCompletedPackets
+	h.evth[evt.EncryptionChangeCode] = h.handleEncryptionChange
 
 	h.subh[evt.LEAdvertisingReportSubCode] = h.handleLEAdvertisingReport
 	h.subh[evt.LEConnectionCompleteSubCode] = h.handleLEConnectionComplete
 	h.subh[evt.LEConnectionUpdateCompleteSubCode] = h.handleLEConnectionUpdateComplete
 	h.subh[evt.LELongTermKeyRequestSubCode] = h.handleLELongTermKeyRequest
-	// evt.EncryptionChangeCode:                     todo),
+	h.subh[evt.EncryptionChangeCode] = h.handleUnhandled
```
Comment (donaldschen) [inline linux/hci/hci.go]: change to: `h.handleEncryptionChange(...)`, or remove

File: linux/hci/hci.go [hunk 3]
```diff
@@ -642,6 +648,10 @@ func (h *HCI) handleDisconnectionComplete(b []byte) error {
 	return nil
 }
 
+func (h *HCI) handleEncryptionChange(b []byte) error {
```
Comment (donaldschen) [inline linux/hci/hci.go:670]: use or remove

File: linux/hci/mgmt/socket.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
+package mgmt
```
Comment (donaldschen) [inline linux/hci/mgmt/socket.go]: lets delete this since it doesn't work with hci

File: linux/hci/smp/context.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import (
+	"bytes"
+	"crypto"
+	"encoding/hex"
+	"fmt"
+	"github.com/go-ble/ble/linux/hci"
+)
+
+type pairingContext struct {
+	request hci.SmpConfig
```
Comment (donaldschen) [inline linux/hci/smp/context.go:12]: would it be worth changing to have 2 nested structs to hold the legacy vs secure connections params?

File: linux/hci/smp/manager.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+//todo: remove bond manager from input parameters?
+func NewSmpManager(config hci.SmpConfig, bm hci.BondManager) *manager {
+	p := &pairingContext{request: config}
+	m := &manager{config:config, pairing:p, bondManager: bm}
+	t := NewSmpTransport(p, bm, m, nil)
+	m.t = t
+	return m
+}
+
+func (m *manager) SetConfg(config hci.SmpConfig) {
```
Comment (donaldschen) [inline linux/hci/smp/manager.go]: typo `SetConfig(...)`

File: linux/hci/smp/manager.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	m.pairing.remoteAddr = swapBuf(remoteAddr)
+	m.pairing.remoteAddrType = remoteAddrType
+
+	m.t.pairing = m.pairing
+
+	fmt.Printf("inited pairing ctx: %+v\n", m.t.pairing)
+}
+
+func (m *manager) Handle(in []byte) error {
+	fmt.Printf("enter handleSMP ====================\n")
```
Comment (donaldschen) [inline linux/hci/smp/manager.go]: remove these printfs?

File: linux/hci/smp/manager.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		}
+
+		return nil
+		// return c.sendSMP(r)
+	}
+
+	fmt.Println("no smp handler...")
+	// FIXME: work around to the lack of SMP implementation - always return non-supported.
+	// C.5.1 Pairing Not Supported by Slave
+	return nil //c.sendSMP([]byte{pairingFailed, 0x05})
```
Comment (donaldschen) [inline linux/hci/smp/manager.go]: I think we should probably re-enable the pairingFailed response

File: linux/device.go [hunk 9]
```diff
@@ -23,14 +24,16 @@ func NewDeviceWithName(name string, opts ...ble.Option) (*Device, error) {
...
-	dev, err := hci.NewHCI(opts...)
+	dev, err := hci.NewHCI(smp2.NewSmpFactory(nil), opts...)
 	if err != nil {
 		return nil, errors.Wrap(err, "can't create hci")
 	}
 	if err = dev.Init(); err != nil {
 		dev.Close()
 		return nil, errors.Wrap(err, "can't init hci")
 	}
+	// _, err = mgmt.NewSocket(1)
```
Comment (donaldschen) [inline linux/device.go]: delete

**Review by donaldschen — COMMENTED** [review]
looks fine just want to remove some stuff from debugging, we need to sort out the behavior when the device does not send a securityRequest but that can be later

File: linux/hci/conn.go (hunk 1 above)
Comment (estutzenberger) [reply linux/hci/conn.go]: yeah, forgot about that, good point

File: linux/hci/hci.go (hunk 2 above)
Comment (estutzenberger) [reply linux/hci/hci.go]: fixed

File: linux/hci/hci.go (hunk 3 above)
Comment (estutzenberger) [reply linux/hci/hci.go:670]: fixed

File: linux/hci/smp/context.go (hunk 5 above)
Comment (estutzenberger) [reply linux/hci/smp/context.go:12]: I think that would be a useful thing but it was hard to get it right (I tried that initially). definitely a future improvement to make

File: linux/hci/smp/manager.go (hunk 7 above)
Comment (estutzenberger) [reply linux/hci/smp/manager.go]: remove a bunch of printfs

File: linux/hci/smp/manager.go (hunk 8 above)
Comment (estutzenberger) [reply linux/hci/smp/manager.go]: done

File: linux/hci/mgmt/socket.go (hunk 4 above)
Comment (estutzenberger) [reply linux/hci/mgmt/socket.go]: done

**Pull request merged**